first cut at zinger rewrite as a kafka syncer (#17)
brimdata/zinger: README.md

## Quick Start

For built-in help, run
```
zinger -h
```
Make sure your config files are setup for the Kafka cluster
and schema registry (see below), then run some tests.

List schemas in the registry:
```
zinger ls
```
Create a topic called `MyTopic` with one partition using your Kafka admin tools,
then post some data to a topic:
```
echo '{s:"hello,world"}' | zinger produce -topic MyTopic -
```
See the record you created:
```
zinger consume -topic MyTopic
```
> Hit Ctrl-C to interrupt `zinger consume` as it will wait indefinitely
> for data to arrive on the topic.

In another shell, run a Zed lake service:
```
mkdir scratch
cd scratch
zed lake serve
```
Now, sync data from Kafka to a Zed lake:
```
zapi create -orderby kafka.offset:desc PoolA
zinger sync from -topic MyTopic -pool PoolA
```
See the data in the Zed pool:
```
zapi query "from PoolA"
```
Next, create a topic called `MyTarget` with one partition using your Kafka admin tools,
sync data from a Zed pool back to Kafka, and check that it made it:
```
zinger sync to -topic MyTarget -pool PoolA
zinger consume -topic MyTarget
```
Finally, try out shaping.  Put a Zed script in `shape.zed`, e.g.,
```
cat shape.zed

> mymeta.foo:=string(2*(kafka.offset+1))

Author: I'm also going to improve this silly example...

```
And shape the record from `MyTopic` into a new `PoolB`:
```
zapi create -orderby kafka.offset:desc PoolB
zinger sync from -topic MyTopic -pool PoolB -shaper shape.zed
zapi query "from PoolB"
```

## Configuration

To configure `zinger` to talk to a Kafka cluster and a schema registry,
you must create two files in `$HOME/.zinger`:

> [`kafka_config.json`](kafka_config.json) and
> [`schema_registry.json`](schema_registry.json).

Reviewer: Maybe change `kafka_config.json` to `kafka.json` for consistency with `schema_registry.json`?
```suggestion
[`kafka_config.json`](kafka_config.json) and
[`schema_registry.json`](schema_registry.json).
```

# zinger

> zinger` is a connector between [Kafka](https://kafka.apache.org/) and

Reviewer: ```suggestion
`zinger` is a connector between [Kafka](https://kafka.apache.org/) and
```

An arbitrary Zed script can be applied to the Zed records in either direction.

The Zed pool used by `zinger` must have its pool key set to `kafka.offset` in
descending order.  `zinger` will detect and report an error if syncing
is attempted using a pool without this configuration.

Each Kafka topic must have a single partition as the system relies upon
the offset to indicate the FIFO order of all records.

### Sync From

`zinger sync from` formats records received from Kafka using the Zed envelope
```
{
  kafka: {topic:string,partition:int64,offset:int64,input_offset:int64},
  key: {...}
  value: {...}
}
```

> where the `key` and `value` fields represent the key/value data pulled from

Reviewer: ```suggestion
where the `key` and `value` fields represent the key and value data pulled from
```

Author: I like key/vaule because it emphasizes the Kafka data model... changing to "key/value data pair"

Kafka and transcoded from Avro to Zed.

If a Zed script is provided, it is applied to each such record before
syncing the data to the Zed pool.  While the script has access to the

> meta-data in the `kafka`, it should not modify these values as this

Reviewer: ```suggestion
metadata in the `kafka`, it should not modify these values as this
```

shaping the logs for a target database schema,
and replicating the shaped CDC logs to a Kafka database
sink connector using `sync to`.

Debezium recommends using a single Kakfa topic for database table.
In this same way, we can scale out the Zed lake and `zinger` processes.

It might be desirable to sync multiple downstream databases with different
schemas to a single upstream database with a unified schema.  This can be
accomplished by having `sync from` read from multiple Kafka topics in parallel
(e,g., reading multiple table formats from different downstream databases),
shape each downstream table accordingly, and store the shaped data in the
unified pool.

A Zed script to shape different schemas to a unified schema is as simple
as a switch statement on the name field of the inbound Kafka topic, e.g.,
```
switch kafka.topic (
  "legacy-oracle-1" => ... ;
  "legacy-oracle-2" => ... ;
  "legacy-mysql-1" => ... ;
  default => ... ;
)
```

> Note that `zinger sync from` does not currently support multiplexing multiple
> inbound topics, but support for this is straightforward and we will add it soon.
>

> > We also need to adapt `sync from` so it updates te consumer commit offsets,

Reviewer 2: ```suggestion
> We also need to adapt `sync from` so it updates the consumer commit offsets,
```

> > allowing aggressive kafka retention policies to drop data that has been

Reviewer: ```suggestion
> allowing aggressive Kafka retention policies to drop data that has been
```
